source/credentials_provider_sts_web_identity.c

```diff
@@ -0,0 +1,1120 @@
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "Failed to resolve token file path during sts web identity provider initialization.")
+        goto on_finish;
+    }
+
+    struct aws_byte_cursor token_file_path_cursor = aws_byte_cursor_from_string(token_file_path);
+    if (aws_byte_buf_init_copy_from_cursor(&parameter->token_file_path, allocator, token_file_path_cursor)) {
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "Failed to resolve token file path during sts web identity provider initialization.")
+        goto on_finish;
+    }
+
+    /* determine role_session_name */
+    if (role_session_name && role_session_name->len) {
+        struct aws_byte_cursor role_session_name_cursor = aws_byte_cursor_from_string(role_session_name);
+        if (aws_byte_buf_init_copy_from_cursor(&parameter->role_session_name, allocator, role_session_name_cursor)) {
+            AWS_LOGF_ERROR(
+                AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+                "Failed to resolve role session name during sts web identity provider initialization: %s",
+                aws_error_str(aws_last_error()));
+            goto on_finish;
+        }
+    } else {
+        struct aws_uuid uuid;
+
+        if (aws_uuid_init(&uuid)) {
+            AWS_LOGF_ERROR(
+                AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+                "Failed to load an uuid during sts web identity provider initialization: %s",
+                aws_error_str(aws_last_error()));
+            goto on_finish;
+        }
+
+        char uuid_str[AWS_UUID_STR_LEN] = {0};
+        struct aws_byte_buf uuid_buf = aws_byte_buf_from_array(uuid_str, sizeof(uuid_str));
+        uuid_buf.len = 0;
+        if (aws_uuid_to_str(&uuid, &uuid_buf)) {
+            AWS_LOGF_ERROR(
+                AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+                "Failed to stringify uuid for role session name during sts web identity provider initialization: %s",
+                aws_error_str(aws_last_error()));
+            goto on_finish;
+        }
+        if (aws_byte_buf_init_copy(&parameter->role_session_name, allocator, &uuid_buf)) {
+            AWS_LOGF_ERROR(
+                AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+                "Failed to generate role session name during sts web identity provider initialization: %s",
+                aws_error_str(aws_last_error()));
+            goto on_finish;
+        }
+    }
+
+    AWS_LOGF_DEBUG(
+        AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+        "Successfully loaded all required parameters for sts web identity credentials provider.")
+    success = true;
+
+on_finish:
+    aws_string_destroy(region);
+    aws_string_destroy(role_arn);
+    aws_string_destroy(role_session_name);
+    aws_string_destroy(token_file_path);
+    aws_string_destroy(profile_name);
+    aws_profile_collection_destroy(config_profile);
+    if (!success) {
+        aws_mem_release(allocator, parameter);
```

Comment:
Won't this leak the bufs inside parameter if they've been partially initialized?  Looks like there's a destroy right below here you could use if moved.

Reply:
good point